dipy/segment/mrf.pyx

```diff
@@ -0,0 +1,657 @@
+#!python
+#cython: boundscheck=False
+#cython: wraparound=False
+#cython: cdivision=True
+import numpy as np
+from dipy.segment.mask import applymask
+from dipy.core.ndindex import ndindex
+from dipy.sims.voxel import add_noise
+cimport cython
+cimport numpy as cnp
+cdef extern from "dpy_math.h" nogil:
+    cdef double NPY_PI
+    cdef double NPY_INFINITY
+    double sqrt(double)
+    double log(double)
+    double exp(double)
+    double fabs(double)
+
+
+class ConstantObservationModel(object):
+    
+    r""" Image segmentation with constant models.
+    Image segmentation by maximum a-posteriori estimation of HMRF model
+    with the Potts/Ising potential
+    """
+
+    def __init__(self):
+        
+        r""" Initializes an instance of the ConstantObservationModel class
+        """
+        pass
+
+
+    def initialize_param_uniform(self, image, nclasses):
+        
+        r""" Initializes the means and variances uniformly
+
+        The means are initialized uniformly along the dynamic range of `image`.
+        The variances are set to 1 for all classes
+
+        Parameters
+        ----------
+        image : array, shape(X, Y, Z)
+        nclasses : int, number of desired classes
+
+        Returns
+        -------
+        mu : array, shape(nclasses,)
+        sigmasq : array, shape(nlasses,)
+        """
+
+        cdef:
+            double[:] mu = np.empty((nclasses,), dtype=np.float64)
+            double[:] sigmasq = np.empty((nclasses,), dtype=np.float64)
+
+        _initialize_param_uniform(image, mu, sigmasq)
+        return np.array(mu), np.array(sigmasq)
+
+
+    def seg_stats(self, input_image, seg_image, nclass):
+        
+        r""" Mean and standard variation for N desired  tissue classes
+
+        Parameters
+        ----------
+        input_image : ndarray, gray level T1 image
+        seg_image : ndarray, initital segmentation image
+        nclass : int, number of classes (3 in most cases)
+
+        Returns
+        -------
+        mu, std, var : ndarrays of dimensions 1x3
+                        Mean, standard deviation and variance for every class
+
+        """
+        mu = np.zeros(nclass)
+        std = np.zeros(nclass)
+        var = np.zeros(nclass)
+
+        for i in range(0, nclass):
+
+            H = input_image[seg_image == i]
+
+            mu[i] = np.mean(H, -1)
+            std[i] = np.std(H, -1)
+            var[i] = np.var(H, -1)
+
+        return mu, std, var
+
+
+    def negloglikelihood(self, image, mu, sigmasq, nclasses):

```

Comment:
it is also much more common to define all mathematicla functions using sigma as an input, instead of using directly sigma**2. It would be less confusing to stick to this convention (or else we could let's say set tau = sigma**2 and forget about the square everywhere).